core/drop_test.go

```diff
@@ -0,0 +1,255 @@
+					},
+
+					&mockTransport{multicastFn: c.gossip},
+				)
+			}
+		},
+	)
+
+	if err := cluster.progressToHeight(5*time.Second, 5); err != nil {
+		t.Fatal("cannot progress to height: err=", err)
+	}
+
+	cluster.stopN(int(cluster.maxFaulty()) + 1)
+
+	assert.Error(t, cluster.progressToHeight(2*time.Second, 10))
+}
+
+/*	HELPERS */
+
+var (
+	validProposal      = []byte("valid proposal")
+	validProposalHash  = []byte("valid proposal hash")
+	validCommittedSeal = []byte("valid committed seal")
+)
+
+func isValidProposal(newProposal []byte) bool {
+	return bytes.Equal(newProposal, validProposal)
+}
+
+func buildValidProposal(_ uint64) []byte {
+	return validProposal
+}
+
+func isValidProposalHash(proposal, proposalHash []byte) bool {
+	return bytes.Equal(
+		proposal,
+		validProposal,
+	) && bytes.Equal(
+		proposalHash,
+		validProposalHash,
+	)
+}
+
+type node struct {
+	core    *IBFT
+	address []byte
+	offline bool
+}
+
+func (n *node) addr() []byte {
+	return n.address
+}
+
+func (n *node) buildPrePrepare(
+	proposal []byte,
+	certificate *proto.RoundChangeCertificate,
+	view *proto.View,
+) *proto.Message {
+	return buildBasicPreprepareMessage(
+		proposal,
+		validProposalHash,
+		certificate,
+		n.address,
+		view,
+	)
+}
+
+func (n *node) buildPrepare(
+	proposal []byte,
+	view *proto.View,
+) *proto.Message {
+	return buildBasicPrepareMessage(
+		validProposalHash,
+		n.address,
+		view,
+	)
+}
+
+func (n *node) buildCommit(
+	proposal []byte,
+	view *proto.View,
+) *proto.Message {
+	return buildBasicCommitMessage(
+		validProposalHash,
+		validCommittedSeal,
+		n.address,
+		view,
+	)
+}
+
+func (n *node) buildRoundChange(
+	proposal []byte,
+	certificate *proto.PreparedCertificate,
+	view *proto.View,
+) *proto.Message {
+	return buildBasicRoundChangeMessage(
+		proposal,
+		certificate,
+		view,
+		n.address,
+	)
+}
+
+type cluster struct {
+	nodes []*node
+	wg    sync.WaitGroup
+
+	latestHeight uint64
+}
+
+func newCluster(num uint64, init func(*cluster)) *cluster {
+	c := &cluster{
+		wg:           sync.WaitGroup{},
+		latestHeight: 0,
+	}
+
+	for _, addr := range generateNodeAddresses(num) {
+		c.nodes = append(c.nodes,
+			&node{address: addr},
+		)
+	}
+
+	init(c)
+
+	return c
+}
+
+func (c *cluster) runSequence(ctx context.Context, height uint64) {
+	for _, n := range c.nodes {
+		c.wg.Add(1)
+
+		go func(n *node) {
+			defer c.wg.Done()
+
+			if n.offline {
+				return
+			}
+
+			n.core.RunSequence(ctx, height)
+		}(n)
+	}
+
+	c.wg.Wait()
+	c.latestHeight = height
+}
+
+func (c *cluster) progressToHeight(timeout time.Duration, height uint64) error {
+	var (
+		ok          = make(chan struct{})
+		ctx, cancel = context.WithTimeout(context.Background(), timeout)
+	)
+
+	defer cancel()
+
+	go func() {
+		for current := c.latestHeight; current < height; current++ {
+			c.runSequence(ctx, current)
+		}
+
+		ok <- struct{}{}
+	}()
+
+	select {
+	case <-ok:
+		//	height reached
+	case <-ctx.Done():
+		return errors.New("timeout")
```

Comment:
Nitick: you can probably extract this error out into a constant

Reply:
There's only one for the time being. I might add more depending on the use cases